feat(worker): pipeline block building

omerfirmak opened the pull request:
This pipelines tracing and CCC checks, aiming to reduce the effective transaction processing delay.
Currently txn processing delay is roughly `tracing_time + ccc_time`. With a pipelined approach, this delay becomes `max(tracing_time, ccc_time)`. Considering that `tracing_time` and `ccc_time` is pretty similar for an average mainnet transaction, this should cut our block building time almost in half.

0xmountaintop commented:
I'm worried that this PR will make "sync (future) l2geth upstream" difficult.
We may want to do a benchmark and see whether it's worth it.

omerfirmak commented:
> I'm worried that this PR will make "sync (future) l2geth upstream" difficult. We may want to do a benchmark and see whether it's worth it.

I believe we can just call this version of the `worker.go` something like `scroll_worker.go`and don't bother merging it with the upstream changes. If there is anything in upstream that's worth incorporating, we can do that later.

omerfirmak commented:
ran some benchmarks against the recent mainnet blocks;
max speed up is 1.94x while average is 1.57x; these should get better as l2geth gets better at creating bigger blocks.

0xmountaintop commented:
> > I'm worried that this PR will make "sync (future) l2geth upstream" difficult. We may want to do a benchmark and see whether it's worth it.
> 
> I believe we can just call this version of the `worker.go` something like `scroll_worker.go`and don't bother merging it with the upstream changes. If there is anything in upstream that's worth incorporating, we can do that later.

great, can we do it this style in this PR?

omerfirmak commented:
> > > I'm worried that this PR will make "sync (future) l2geth upstream" difficult. We may want to do a benchmark and see whether it's worth it.
> > 
> > 
> > I believe we can just call this version of the `worker.go` something like `scroll_worker.go`and don't bother merging it with the upstream changes. If there is anything in upstream that's worth incorporating, we can do that later.
> 
> great, can we do it this style in this PR?

Sure, will do :+1:

omerfirmak commented:
> these should get better as l2geth gets better at creating bigger blocks

Re-ran the test against the blocks from this morning

```
INFO [05-07|10:53:41.568] PStats                                   lifetime=919.855626ms apply=881.202034ms apply_idle=0s apply_stall=528.638681ms ccc=737.891694ms ccc_idle=161.223213ms speedup=1.760 min=1.366 max=1.939 avg=1.726
INFO [05-07|10:53:42.544] PStats                                   lifetime=923.765538ms apply=923.486902ms apply_idle=0s apply_stall=401.590975ms ccc=646.22426ms  ccc_idle=261.266619ms speedup=1.699 min=1.366 max=1.939 avg=1.726
INFO [05-07|10:53:43.602] PStats                                   lifetime=986.193733ms apply=975.360039ms apply_idle=0s apply_stall=477.601506ms ccc=652.291066ms ccc_idle=284.070849ms speedup=1.650 min=1.366 max=1.939 avg=1.726
```

avg speed up is 1.726 now.

0xmountaintop commented:
> > > > I'm worried that this PR will make "sync (future) l2geth upstream" difficult. We may want to do a benchmark and see whether it's worth it.
> > > 
> > > 
> > > I believe we can just call this version of the `worker.go` something like `scroll_worker.go`and don't bother merging it with the upstream changes. If there is anything in upstream that's worth incorporating, we can do that later.
> > 
> > 
> > great, can we do it this style in this PR?
> 
> Sure, will do 👍

if we keep `miner/worker.go` (instead of deleting it), and only add a `/*` and a `*/` to comment out all the lines below `package miner`, then we will only have 2 lines diff (instead of the whold file diff).
This way it will be easier to track future changes in the upstream (or we can also look into PRs in upstream 1 by 1 though).

0xmountaintop commented on miner/scroll_worker.go:
```diff
@@ -0,0 +1,958 @@
+		case head := <-w.chainHeadCh:
+			clearPending(head.Block.NumberU64())
+			timestamp = time.Now().Unix()
+			commit(true)
```
I notice that we remove the timer trigger for recommit here, can you explain the trade-off?

omerfirmak replied:
recommit is only relevant for "mining", where the node is constantly rebuilding a block for the next height to find the most profitable one. Sequencers don't have that kind of behaviour, block for a given height is built only once and committed to.

omerfirmak committed:
feat(worker): pipeline block building

omerfirmak commented:
> > > > > I'm worried that this PR will make "sync (future) l2geth upstream" difficult. We may want to do a benchmark and see whether it's worth it.
> > > > 
> > > > 
> > > > I believe we can just call this version of the `worker.go` something like `scroll_worker.go`and don't bother merging it with the upstream changes. If there is anything in upstream that's worth incorporating, we can do that later.
> > > 
> > > 
> > > great, can we do it this style in this PR?
> > 
> > 
> > Sure, will do 👍
> 
> if we keep `miner/worker.go` (instead of deleting it), and only add a `/*` and a `*/` to comment out all the lines below `package miner`, then we will only have 2 lines diff (instead of the whold file diff). This way it will be easier to track future changes in the upstream (or we can also look into PRs in upstream 1 by 1 though).

Done.

omerfirmak commented on rollup/pipeline/pipeline.go:
```diff
@@ -0,0 +1,407 @@
+					p.nextL1MsgIndex = tx.AsL1MessageTx().QueueIndex + 1
+				}
+
+				stallStart := time.Now()
```
We can consider moving the commit phase to here as well. Instead of stalling to wait for CCC stage, we can calculate the state commitment-per-txn here, effectively making commit phase entirely free.

omerfirmak committed:
tidy up metrics

0xmountaintop commented:
Isn't it still processing txs sequentially? still trying to understand where the speed gain is from

omerfirmak commented:
> Isn't it still processing txs sequentially? still trying to understand where the speed gain is from

Improvement comes from the fact that we can run tracing and CCC check in parallel. So for example; when we are tracing txn(i), we can run CCC check on txn(i-1) in parallel.

0xmountaintop commented on rollup/pipeline/pipeline.go:
```diff
@@ -0,0 +1,407 @@
+		}
+
+		switch {
+		case err == nil, errors.Is(err, core.ErrNonceTooLow):
+			txs.Shift()
+		default:
+			if errors.Is(err, ErrApplyStageDone) || onFailingTxn(p.txs.Len(), tx, err) {
+				close(p.TxnQueue)
+				p.TxnQueue = nil
+				return nil
+			}
+			txs.Pop()
```
the shift/pop behavior here is different from the original logic, especially for L1MsgTx

0xmountaintop replied:
if you believe this behavior should be changed, can you explain?

omerfirmak replied:
AFAICS, the behaviour change is very minor and only relevant for L1 messages. I will add a special casing for L1 messages here.

omerfirmak replied:
done here https://github.com/scroll-tech/go-ethereum/pull/735/commits/19dc956bf072b77713edf18c287f701a1c96a218

0xmountaintop commented on miner/scroll_worker.go:
```diff
@@ -0,0 +1,943 @@
+	}
+}
+
+func (w *worker) onTxFailingInPipeline(txIndex int, tx *types.Transaction, err error) bool {
```
what does this return value indicate? it seems always returning false

omerfirmak replied:
It allows caller to terminate the pipeline early when a txn fails apply stage. https://github.com/scroll-tech/go-ethereum/pull/735/files#diff-3236476391a7611535994ed7b11821c5e7b6cf6056e0a49c498dbfb1e249f43dR129

I will document it better.

0xmountaintop commented on miner/scroll_worker.go:
```diff
@@ -0,0 +1,943 @@
+		case res.OverflowingTx.IsL1MessageTx() &&
+			errors.Is(res.CCCErr, circuitcapacitychecker.ErrUnknown):
+			l1TxCccUnknownErrCounter.Inc(1)
+		case !res.OverflowingTx.IsL1MessageTx() &&
+			errors.Is(res.CCCErr, circuitcapacitychecker.ErrUnknown):
+			l2TxCccUnknownErrCounter.Inc(1)
```
previously we `RemoveTx` for this case (`circuitcapacitychecker.ErrUnknown` & `!IsL1MessageTx`).

again, if you believe this behavior should be changed, can you explain?

omerfirmak replied:
This switch-case is only for metrics collection.
Transaction is removed from the pool at [this](https://github.com/scroll-tech/go-ethereum/pull/735/files#diff-1825330f8f75827945cb0144a6ab3dc034a3081f48ee96b0a318c5de9950832eR789) line

0xmountaintop replied:
but the code you linked will only run when `res.FinalBlock == nil` (first txn overflowed the circuit)

what if the first several txns are okay, but then a following txn returns `circuitcapacitychecker.ErrUnknown`?

omerfirmak replied:
All CCC errors are treated the same. If a txn fails CCC check, it gets processed as the first txn the next block. If it fails CCC check again as the first txn in the next block, it gets skipped and removed from the mempool. So, all txns get a second chance at going thru the CCC.

0xmountaintop commented on miner/scroll_worker.go:
```diff
@@ -0,0 +1,943 @@
+		writeTrace()
+		l1TxGasLimitExceededCounter.Inc(1)
+
+	case errors.Is(err, core.ErrInsufficientFunds):
```
there's a reason why we did unwrap
(why we used `errors.Is(err, core.ErrInsufficientFunds) || errors.Is(errors.Unwrap(err), core.ErrInsufficientFunds)`)  before:

see https://github.com/scroll-tech/go-ethereum/blob/develop/core/state_transition.go#L256

omerfirmak replied:
`errors.Is` already handles wrapped errors. https://cs.opensource.google/go/go/+/refs/tags/go1.22.3:src/errors/wrap.go;l=29-31

0xmountaintop commented:
the workflow in this PR looks good to me, I'm just worried about the behavior changes on the error handling (pop/shift/removeTx)

omerfirmak committed:
shift failing l1 messages instead of popping them

omerfirmak committed:
Merge remote-tracking branch 'origin/develop' into omerfirmak/pipelined-worker

0xmountaintop commented:
btw, I have a minor suggestion PR on the style: https://github.com/scroll-tech/go-ethereum/pull/749

Thegaram commented on rollup/pipeline/pipeline.go:
```diff
@@ -0,0 +1,412 @@
+	parent   *types.Block
+	start    time.Time
+
+	// accumalators
```
```suggestion
	// accumulators
```

Thegaram commented on rollup/pipeline/pipeline.go:
```diff
@@ -0,0 +1,412 @@
+	gasPool        *core.GasPool
+
+	// com channels
+	TxnQueue         chan *types.Transaction
```
Is it necessary to export this field?

omerfirmak replied:
Caller needs to be able to close it and block on it. We can put `TxnQueue` and `ResultCh`behind accessor functions, but I just made them public for convenience. Let me know if you think they are better left private.

Thegaram replied:
Is this still true? The only place `scroll_worker` uses it is this check:
```go
w.currentPipeline.TxnQueue != nil
```

omerfirmak replied:
made private https://github.com/scroll-tech/go-ethereum/pull/735/commits/f44d1a9e5712f356a7c72505db2d7f66737d8c71

Thegaram commented on rollup/pipeline/pipeline.go:
```diff
@@ -0,0 +1,412 @@
+		nextL1MsgIndex: *rawdb.ReadFirstQueueIndexNotInL2Block(chain.Database(), parent.Hash()),
+		Header:         *header,
+		ccc:            ccc,
+		coinbase:       coinbase,
```
This is not really used (see `NewEVMBlockContext`). I think it was only kept for compatibility with old tests. So we can just pass `nil` to `ApplyTransaction ` and add a comment about `FeeVaultEnabled`.

omerfirmak replied:
removed in https://github.com/scroll-tech/go-ethereum/pull/735/commits/ad086f9b7dcbda95565d91f3d6943dbe07760c50. removing it from worker as well required to touching up a lot more files.

Thegaram commented on rollup/pipeline/pipeline.go:
```diff
@@ -0,0 +1,412 @@
+			}
+
+			// If we have collected enough transactions then we're done
+			// Originally we only limit l2txs count, but now strictly limit total txs number.
```
> `// Originally we only limit l2txs count, but now strictly limit total txs number.`

Why?

omerfirmak replied:
That is not a new comment, I copied it over from the original worker.

Thegaram commented on rollup/pipeline/pipeline.go:
```diff
@@ -0,0 +1,412 @@
+					return
+				}
+				deadlineReached = true
+				deadline = time.Now().Add(time.Hour)
```
Why do we add 1 hour here?

omerfirmak replied:
It is hacky way of making sure that deadline case is not triggered repeatedly after the deadline is hit.

Thegaram replied:
Makes sense. Maybe it's better to add a comment about this.

omerfirmak replied:
done https://github.com/scroll-tech/go-ethereum/pull/735/commits/9d1147ad29fff58c90b7ad63fe0bad4822aca24a

Thegaram commented on rollup/pipeline/pipeline.go:
```diff
@@ -0,0 +1,412 @@
+				cccStart := time.Now()
+				var accRows *types.RowConsumption
+				var err error
+				if increment != nil {
```
Under what circumstances will this be `nil`?

omerfirmak replied:
When apply stage closes the increment channel.

Thegaram replied:
Actually what I meant to ask is: Is it possible that `increment.LastTrace == nil` (next line)?

omerfirmak replied:
I think it is not possible at the moment, will remove.

omerfirmak replied:
done https://github.com/scroll-tech/go-ethereum/pull/735/commits/8da0960ec30bb7042b0aaf8b9b6815c166dcd7b8

Thegaram commented on rollup/pipeline/pipeline.go:
```diff
@@ -0,0 +1,412 @@
+	}
+}
+
+type PendingBlockIncrement struct {
```
This does not feel an "increment" to me, since `Header`, `State`, `Txs` etc contain all changes up to and including the latest tx processed (not just the incremental change). Not sure what might be a better name though.

omerfirmak replied:
How about "progress"?

Thegaram replied:
How about `BlockCandidate` or `NewBlockCandidate`?

omerfirmak replied:
`BlockCandidate`sounds good.

omerfirmak replied:
done https://github.com/scroll-tech/go-ethereum/pull/735/commits/a53eb879b83847b86e021509a39d10321c9d38db

Thegaram reviewed (commented):
Review wip, leaving some initial comments and questions.

Thegaram commented on rollup/pipeline/pipeline.go:
```diff
@@ -0,0 +1,412 @@
+			return result
+		}
+
+		switch {
```
Shouldn't we handle `ErrGasLimitReached` here?

omerfirmak replied:
what kind of special casing do you think it needs?

Thegaram replied:
The current flow is: When we encounter `ErrGasLimitReached`, we propagate it to `onTxFailingInPipeline`, which does not terminate the pipeline (it does not return `true`). Is that right?

Or is the thinking that this will not happen because of the `if p.gasPool.Gas() < params.TxGas {` check?

omerfirmak replied:
> When we encounter ErrGasLimitReached, we propagate it to onTxFailingInPipeline, which does not terminate the pipeline (it does not return true). Is that right?

Correct, simply because we might be able to find another transaction that we can include in the block.

> Or is the thinking that this will not happen because of the if p.gasPool.Gas() < params.TxGas { check?

ErrGasLimitReached errors will happen but we don't need to finalize block building immediately when we see one. When the apply stage is certain that it cannot put any more txns in the block, it will terminate and force the next stage to seal the block immediately.

Thegaram commented on rollup/pipeline/pipeline.go:
```diff
@@ -0,0 +1,412 @@
+					lastAccRows = accRows
+				}
+
+				// immediately close the block if deadline reached or apply staged is done
```
```suggestion
				// immediately close the block if deadline reached or apply stage is done
```

Thegaram commented on miner/scroll_worker.go:
```diff
@@ -0,0 +1,943 @@
+			return
+		}
+
+		if result := w.currentPipeline.TryPushTxns(txs, w.onTxFailingInPipeline); result != nil {
```
Just to make sure I understand: `TryPushTxns` will wait for the trace result but not for the CCC result, right?

Also, by the time `startNewPipeline` returns, it's likely that the block is already finished, right?

omerfirmak replied:
> TryPushTxns will wait for the trace result but not for the CCC result, right

Correct, but if CCC result happens to come while it is waiting for the trace result, it will return the CCC result to caller.

> Also, by the time startNewPipeline returns, it's likely that the block is already finished, right?

If there are enough txns in the pool to trigger the CCC, certainly. If not, we will wait until the deadline and keep appending the txns that enter the pool after `startNewPipeline`returns.

Thegaram commented on miner/scroll_worker.go:
```diff
@@ -0,0 +1,943 @@
+	}
+
+	if res == nil || res.FinalBlock == nil {
+		w.startNewPipeline(time.Now().Unix())
```
If `startNewPipeline` calls `handleCCCResult`, and then here we call `startNewPipeline`, will this second execution succeed (as opposed to the first one)?

omerfirmak replied:
This case is for when the first txn fails the CC check. [The lines](https://github.com/scroll-tech/go-ethereum/pull/735/files#diff-1825330f8f75827945cb0144a6ab3dc034a3081f48ee96b0a318c5de9950832eR773-R791) above skips that txn. So the second time we create a pipeline hopefully we won't be getting another txn that we need to skip and be able to create a new block.

omerfirmak committed:
remove unneccassary if

omerfirmak committed:
comment on the updated deadline

omerfirmak committed:
rename increment to candidate

omerfirmak committed:
Update rollup/pipeline/pipeline.go

Co-authored-by: Péter Garamvölgyi <peter@scroll.io>

omerfirmak committed:
properly type channels to indicate directions

omerfirmak committed:
Merge branch 'develop' into omerfirmak/pipelined-worker

omerfirmak committed:
chore: auto version bump [bot]

omerfirmak committed:
remove coinbase config from pipeline

omerfirmak committed:
use PendingWithMax in scroll_worker

omerfirmak committed:
add committed gas counter metric

Thegaram commented on miner/scroll_worker.go:
```diff
@@ -0,0 +1,945 @@
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package miner
```
As I recall you said that the pipeline is also used in follower nodes, right? Which part is that?

omerfirmak replied:
I initially thought that was a good idea, but followers don't really benefit from pipeline. They already trace the block in a parallelized way.

Thegaram commented on rollup/pipeline/pipeline.go:
```diff
@@ -0,0 +1,408 @@
+		chain:          chain,
+		vmConfig:       vmConfig,
+		parent:         parent,
+		nextL1MsgIndex: *rawdb.ReadFirstQueueIndexNotInL2Block(chain.Database(), parent.Hash()),
```
Better to handle the `nil` case here.

omerfirmak replied:
done https://github.com/scroll-tech/go-ethereum/commit/cccfc1738df7114d7daf87328340dc678e61572e

Thegaram replied:
We should return error on the `dbIndex == nil` case or the produced block will be incorrect.

omerfirmak replied:
done [b81a59c83](https://github.com/scroll-tech/go-ethereum/pull/735/commits/b81a59c83e8bd186b4d73a579c0e907fc97a021f)

Thegaram commented on rollup/pipeline/pipeline.go:
```diff
@@ -0,0 +1,408 @@
+					lastTxn := candidate.Txs[candidate.Txs.Len()-1]
+					cccTimer.UpdateSince(cccStart)
+					if err != nil {
+						resultCh <- &Result{
```
Where do we handle the case where it's the first tx that's overflowing?

omerfirmak replied:
https://github.com/scroll-tech/go-ethereum/pull/735/files#diff-1825330f8f75827945cb0144a6ab3dc034a3081f48ee96b0a318c5de9950832eR762

Thegaram commented on rollup/pipeline/pipeline.go:
```diff
@@ -0,0 +1,408 @@
+				return nil
+			}
+
+			if tx.IsL1MessageTx() {
+				txs.Shift()
+			} else {
+				txs.Pop()
+			}
```
There's some subtle change in behaviour here. E.g. in the "strange error" branch in the original code, we'd `txs.Shift()` for both types. In this case it won't break anything, just want to check if you reviewed all the different error cases.

omerfirmak replied:
Yeah, I went thru all of them. `txs.Shift()` is only relevant for L1 messages and ErrNonceTooLow errors.  

> .g. in the "strange error" branch in the original code, we'd txs.Shift() for both types

for example, in this case we will end up needing to Shift() all pending txns of that account anyways. Because once a txn fails, the rest will too because of their nonce.

Thegaram commented on rollup/pipeline/pipeline.go:
```diff
@@ -0,0 +1,408 @@
+	select {
+	case err, valid := <-p.applyStageRespCh:
+		if !valid {
+			return nil, ErrApplyStageDone
```
In these early return cases (gas limit, tx count, etc) will we only get the result (from `p.ResultCh`) after the timeout? Will this introduce an unnecessary delay?

omerfirmak replied:
> will we only get the result (from p.ResultCh) after the timeout?

Nope, `traceAndApplyStage`will close its downstream `newCandidateCh` which will signal the next stage to immediately seal the block.

Thegaram commented on rollup/pipeline/pipeline.go:
```diff
@@ -0,0 +1,408 @@
+	FinalBlock *BlockCandidate
+}
+
+func (p *Pipeline) cccStage(candidates <-chan *BlockCandidate, deadline time.Time) <-chan *Result {
```
Will `deadline` replace the `seal block early if we're over time` logic from the original code?

omerfirmak replied:
Yeap, it does.

Thegaram commented on rollup/pipeline/pipeline.go:
```diff
@@ -0,0 +1,408 @@
+		p.beforeTxHook()
+	}
+
+	if p.ccc != nil {
```
Why do we need this check?

omerfirmak replied:
Not needed, will remove.

omerfirmak replied:
done https://github.com/scroll-tech/go-ethereum/pull/735/commits/7cd959c7a2c386abb67eb8010c0c77f897ee24a9

Thegaram commented on miner/scroll_worker.go:
```diff
@@ -0,0 +1,945 @@
+	if recommit < minRecommitInterval {
+		log.Warn("Sanitizing miner recommit interval", "provided", recommit, "updated", minRecommitInterval)
+		recommit = minRecommitInterval
+	}
+
+	worker.wg.Add(4)
```
Please add new code from my recent PR:

```go
// Sanitize account fetch limit.
if worker.config.MaxAccountsNum == 0 {
	log.Warn("Sanitizing miner account fetch limit", "provided", worker.config.MaxAccountsNum, "updated", math.MaxInt)
	worker.config.MaxAccountsNum = math.MaxInt
}
```

omerfirmak replied:
done https://github.com/scroll-tech/go-ethereum/pull/735/commits/9d2f6b1ef4f2bdbf3685c7da8b88c2a5cae46fbf

Thegaram commented on miner/scroll_worker.go:
```diff
@@ -0,0 +1,945 @@
+			atomic.AddInt32(&w.newTxs, int32(len(ev.Txs)))
+
+		case ev := <-w.l1MsgsCh:
+			atomic.AddInt32(&w.newL1Msgs, int32(ev.Count))
```
Seems like this is not used anymore?

omerfirmak replied:
Yeah, but that was the case for the existing worker as well. Will remove regardless.

omerfirmak replied:
https://github.com/scroll-tech/go-ethereum/pull/735/commits/62e71907daf43aaea87432656dbe70c9f2b70a23

Thegaram commented on miner/scroll_worker.go:
```diff
@@ -0,0 +1,945 @@
+	}
+
+	parent := w.chain.CurrentBlock()
+
```
The `parent.Time() >= uint64(timestamp)` check was removed, is that intentional?

omerfirmak replied:
Yes it is redundant, `w.engine.Prepare`sanity checks the timestamp already.

Thegaram commented on miner/scroll_worker.go:
```diff
@@ -0,0 +1,945 @@
+	}
+}
+
+func (w *worker) handleCCCResult(res *pipeline.Result) error {
```
```suggestion
func (w *worker) handlePipelineResult(res *pipeline.Result) error {
```

omerfirmak replied:
https://github.com/scroll-tech/go-ethereum/pull/735/commits/235246de4f625806be5763d659cc8fbb58f250e7

Thegaram commented on miner/scroll_worker.go:
```diff
@@ -0,0 +1,945 @@
+
+			// Store skipped transaction in local db
+			overflowingTrace := res.OverflowingTrace
+			if !w.config.StoreSkippedTxTraces {
+				overflowingTrace = nil
+			}
+			rawdb.WriteSkippedTransaction(w.eth.ChainDb(), res.OverflowingTx, overflowingTrace, res.CCCErr.Error(),
+				w.currentPipeline.Header.Number.Uint64(), nil)
```
We should not write to db when `!w.config.StoreSkippedTxTraces`

omerfirmak replied:
It follows the behaviour of original worker. 

```
				// Store skipped transaction in local db
				if w.config.StoreSkippedTxTraces {
					rawdb.WriteSkippedTransaction(w.eth.ChainDb(), tx, traces, "row consumption overflow", w.current.header.Number.Uint64(), nil)
				} else {
					rawdb.WriteSkippedTransaction(w.eth.ChainDb(), tx, nil, "row consumption overflow", w.current.header.Number.Uint64(), nil)
				}

 
```

Thegaram replied:
Oh makes sense. So we store the skipped tx in both cases, but not the traces.

Thegaram commented on miner/scroll_worker.go:
```diff
@@ -0,0 +1,945 @@
+				w.currentPipeline.Header.Number.Uint64(), nil)
+
+			if overflowingL1MsgTx := res.OverflowingTx.AsL1MessageTx(); overflowingL1MsgTx != nil {
+				rawdb.WriteFirstQueueIndexNotInL2Block(w.eth.ChainDb(), w.currentPipeline.Header.ParentHash, overflowingL1MsgTx.QueueIndex+1)
```
Why write to DB here? Is this the only way to ensure that we'll start with the next l1MsgTx in the next pipeline?

omerfirmak replied:
Yes, correct. Only way that I could come up with to persist that information between pipeline instances.

Thegaram commented on miner/scroll_worker.go:
```diff
@@ -0,0 +1,945 @@
+		commitReasonDeadlineCounter.Inc(1)
+	}
+	commitGasCounter.Inc(int64(res.FinalBlock.Header.GasUsed))
+
```
Is the `// set w.current.accRows for empty-but-not-genesis block` part from the original code handled somewhere?

omerfirmak replied:
No, empty-but-not-genesis blocks are impossible with this implementation.

Thegaram replied:
Okay makes sense. Might be better to leave a comment about this, or some kind of an assertion. Because if we change this code in the future, and it will produce an empty block, it's likely that we'll forget to run CCC for the empty block.

omerfirmak replied:
added here https://github.com/scroll-tech/go-ethereum/pull/735/commits/6436cc64a9b686fd7d8afb7fd4744d80c0c86d98

Thegaram commented on miner/scroll_worker.go:
```diff
@@ -0,0 +1,945 @@
+	switch {
+	case errors.Is(err, core.ErrGasLimitReached) && tx.IsL1MessageTx():
+		// If this block already contains some L1 messages,
+		// terminate here and try again in the next block.
```
But we don't really terminate here, for that we'd need to `return true`

omerfirmak replied:
copy pasta error. fixed https://github.com/scroll-tech/go-ethereum/pull/735/commits/05bbd32e904c8447460d601619779e1c935987a4

omerfirmak committed:
handle nil index in db

omerfirmak committed:
chore: auto version bump [bot]

omerfirmak committed:
Merge remote-tracking branch 'origin/develop' into omerfirmak/pipelined-worker

omerfirmak committed:
sanitize MaxAccountsNum

omerfirmak committed:
remove l1msg subscription

omerfirmak committed:
handlePipelineResult

omerfirmak committed:
fix comment on erroring l1 msg

omerfirmak committed:
unexport TxnQueue

omerfirmak committed:
move db read for l1msgindex to caller site

omerfirmak committed:
empty block note

omerfirmak committed:
remove redundant nil check

omerfirmak commented:
> btw, I have a minor suggestion PR on the style: #749

From what I can see, the most notable change in that PR is moving the local channels to Pipeline scope. I kinda like that they are local to the stage functions because it signifies that the owners of those channels are the respective stages, not the pipeline itself. It makes reasoning about how stages interact with each other easier and disallows accidentally interacting with the channels from outside of the stages.

roylou commented:
I didn't read the code. Yet you add 900 lines of code (scroll_worker.go) and deletes 300 lines of testing code (scroll_worker_test.go). How can we make sure the newly added code are well tested?

omerfirmak commented:
> I didn't read the code. Yet you add 900 lines of code (scroll_worker.go) and deletes 300 lines of testing code (scroll_worker_test.go). How can we make sure the newly added code are well tested?

Deleted tests are for the features that we removed as well. Newly added code directly plugs in to the paths that existing tests cover, so from the callers perspective, behaviour of worker stays the same.
